### PR TITLE
New - Support 'G' and 'gg' movements in multi-line buffers.

### PR DESCRIPTION
**Pull request by springcomp**
As discussed, in [#789](https://github.com/lzybkr/PSReadLine/issues/789) here is a pull request, and accompanying test.

This supports <kbd>G</kbd> and <kbd>gg</kbd> movements to, respectively, move to the end or the beginning of a multiline buffer.

Please, let me know if I need to provide more informations or put some more work into this pull request.

### CODE REVIEW

File: PSReadLine/Render.cs [hunk 1]
```diff
@@ -799,6 +799,40 @@ private int ConvertLineAndColumnToOffset(Point point)
...
+            }
+            return number;
+
+        }
+
+        /// <summary>
+        /// Returns the number of lines in a multi-line buffer
+        /// </summary>
+        /// <returns></returns>
+        private int GetLineCount()
```
Comment (lzybkr) [inline PSReadLine/Render.cs]: This should probably be named `GetLogicalLineCount` so it isn't used accidentally.

When rendering, a logical line may span multiple actual lines (see [PhysicalLineCount](https://github.com/lzybkr/PSReadLine/blob/bc40ae6f6b44e2eb4bd8b57810079bfd89bffa72/PSReadLine/Render.cs#L391)), so it's useful to know which kind of line count it is.

File: PSReadLine/Render.cs [hunk 2]
```diff
@@ -799,6 +799,40 @@ private int ConvertLineAndColumnToOffset(Point point)
...
+        {
+            var count = 1;
+
+            for (int i = 0; i < _buffer.Length; i++)
+            {
+                if (_buffer[i] == '\n')
+                    count++;
```
Comment (lzybkr) [inline PSReadLine/Render.cs:834]: Braces.

File: PSReadLine/Render.cs [hunk 3]
```diff
@@ -799,6 +799,40 @@ private int ConvertLineAndColumnToOffset(Point point)
             return (point.Y == y) ? offset : -1;
         }
 
+        /// <summary>
+        /// Returns the line number under the cursor.
+        /// </summary>
+        /// <returns></returns>
```
Comment (lzybkr) [inline PSReadLine/Render.cs]: Remove empty doc comment.

File: PSReadLine/Render.cs [hunk 4]
```diff
@@ -799,6 +799,40 @@ private int ConvertLineAndColumnToOffset(Point point)
...
+                    number++;
+            }
+            return number;
+
+        }
+
+        /// <summary>
+        /// Returns the number of lines in a multi-line buffer
+        /// </summary>
+        /// <returns></returns>
```
Comment (lzybkr) [inline PSReadLine/Render.cs]: Remove empty doc comment.

File: test/MovementTest.VI.Multiline.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                _.DQuote, "Multiline buffer", _.Enter,
+                " containing an empty line", _.Enter,
+                _.Enter,
+                "and text aligned on the left", _.Enter,
+                _.DQuote,
+                _.Escape, CheckThat(() => AssertCursorTopIs(4)),
+                "gg", CheckThat(() => AssertCursorLeftTopIs(0, 0)),
+                'G', CheckThat(() => AssertCursorLeftTopIs(continuationPrefixLength, 4))
+            ));
+
```
Comment (lzybkr) [inline test/MovementTest.VI.Multiline.cs:41]: Please add a test for an empty buffer and for a single line buffer.

**Review by lzybkr — CHANGES_REQUESTED** [review]
Looks great, I just have some minor feedback to address and then this should be good to merge.

File: PSReadLine/Movement.vi.multiline.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        /// Moves the cursor to the beginning of the first line.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="arg"></param>
+        public void MoveToFirstLine(ConsoleKeyInfo? key = null, object arg = null)
+        {
+            if (!LineIsMultiLine())
+                Ding(key, arg);
+
+            var number = GetCurrentLine();
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs]: All right, I'm good with your suggestion too.

File: PSReadLine/Movement.vi.multiline.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+        /// <summary>
+        /// Moves the cursor to the beginning of the last line.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="arg"></param>
+        public void MoveToLastLine(ConsoleKeyInfo? key = null, object arg = null)
+        {
+            if (!LineIsMultiLine())
+                Ding(key, arg);
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs:43]: Yep.

File: PSReadLine/Movement.vi.multiline.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+            var count = GetLineCount();
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs]: Oki doki.

File: PSReadLine/Movement.vi.multiline.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            var number = GetCurrentLine();
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs]: I get the gist of it...

File: PSReadLine/Render.cs (hunk 2 above)
Comment (springcomp) [reply PSReadLine/Render.cs:834]: I'm starting to get the drift.

File: PSReadLine/Render.cs (hunk 3 above)
Comment (springcomp) [reply PSReadLine/Render.cs]: You mean the empty `returns` clause, or the whole comment ?

File: PSReadLine/Render.cs (hunk 4 above)
Comment (springcomp) [reply PSReadLine/Render.cs]: Ok.

File: test/MovementTest.VI.Multiline.cs (hunk 5 above)
Comment (springcomp) [reply test/MovementTest.VI.Multiline.cs:41]: I will make sure to cover those tests as well, thank you.

File: PSReadLine/Movement.vi.multiline.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        /// <summary>
+        /// Moves the cursor to the beginning of the first line.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="arg"></param>
+        public void MoveToFirstLine(ConsoleKeyInfo? key = null, object arg = null)
+        {
+            if (!LineIsMultiLine())
+                Ding(key, arg);
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs:17]: No worries, I'll make sure to follow the best practices. Sorry about that.

File: PSReadLine/Render.cs (hunk 3 above)
Comment (lzybkr) [reply PSReadLine/Render.cs]: Just the `returns` clause - it's just noise with actual documentation.

**Comment by springcomp** [comment]
I think I have complied with all your requests, thank you for your feedback.
Sorry about the messy commits. I'll make sure to improve this in subsequent pull requests.

File: PSReadLine/Movement.vi.multiline.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+{
+    public partial class PSConsoleReadLine
+    {
+        /// <summary>
+        /// Moves the cursor to the beginning of the first logical line
+        /// of a multi-line buffer.
+        /// </summary>
+        /// <param name="key" />
+        /// <param name="arg" />
+        public void MoveToFirstLogicalLine(ConsoleKeyInfo? key = null, object arg = null)
```
Comment (lzybkr) [inline PSReadLine/Movement.vi.multiline.cs]: Hmm, I hadn't considered the public functions using `LogicalLine`, just the private function.

This function name is exposed, e.g. you'd see it when listing key bindings or in the documentation. I'm thinking `Logical` isn't necessary here.

File: PSReadLine/Movement.vi.multiline.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            _singleton.MoveCursor(position);
+        }
+
+        /// <summary>
+        /// Moves the cursor to the beginning of the last logical logical line.
+        /// of a multi-line buffer.
+        /// </summary>
+        /// <param name="key" />
+        /// <param name="arg" />
+        public void MoveToLastLogicalLine(ConsoleKeyInfo? key = null, object arg = null)
```
Comment (lzybkr) [inline PSReadLine/Movement.vi.multiline.cs]: Same thing here - I think `Logical` isn't necessary on the public api.

**Comment by lzybkr** [comment]
You should also add cases in `GetDisplayGrouping` - these new bindings should get grouped under `KeyHandlerGroup.CursorMovement`.

File: PSReadLine/Movement.vi.multiline.cs (hunk 11 above)
Comment (springcomp) [reply PSReadLine/Movement.vi.multiline.cs]: Good catch. I've reverted the function names as instructed.

**Review by lzybkr — APPROVED** [review]

**Comment by lzybkr** [comment]
Thanks for the contribution!

**Comment by springcomp** [comment]
Thank you for this awesome project.
I'm looking forward to contributing again.
